### DISCUSSION TITLE
Help with `remoteRoot` and `localRoot`

### DISCUSSION THREAD

**Discussion by jdevries3133** [mfussenegger/nvim-dap#596] [Q&A]
Hi,

I'm having a rough time trying to figure out how to debug a containerized node application with nvim-dap and vscode-node2-debug. I created a minimal repo which demonstrates the problem I'm having:

https://github.com/jdevries3133/trivial_example

Basically, source maps don't seem to be working, and breakpoints set through nvim-dap don't work. I was having the same problem in VS Code too, until I realized I needed to set the `remoteRoot` and `localRoot` settings in my `launch.json`. After doing that, VS Code works find, but nvim-dap still doesn't work.

I tried pretty hard to debug this, even cloning the `vscode-node-debug` and `vscode-chrome-debug-core` repositories and throwing some console logs into them to see if the `localRoot` and `remoteRoot` values were being passed through nvim-dap into the debug adapter, and it seemed like things *should* have been working.

Any tips for getting to the root of this problem are very much appreciated!

# Update

Actually, this trivial solution is working now (probably, it always was and I was mistaken). My real use case still isn't working, so there must be a different problem I haven't isolated yet. For context, I'm trying to debug a remix.run application inside a docker container with remote debugging and this is my launch.json for that project:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "address": "localhost",
      "name": "Attach to Remote",
      "port": 9229,
      "request": "attach",
      "skipFiles": ["<node_internals>/**"],
      "type": "node",
      "sourceMaps": true,
      "localRoot": "${workspaceFolder}",
      "remoteRoot": "/app",
      "outFiles": [
        "${workspaceFolder}/build/*.js",
        "${workspaceFolder}/pubilc/build/*.js",
        "!**/node_modules/**"
      ]
    }
  ]
}
```

Anyway, I will keep investigating

# Update 2

OK, now this is so weird 🙃

If I set a breakpoint in the generated code, then continue the program, the debugger will jump to the *actual* source code which maps to that generated code, but I still can't set breakpoints in the source code.

**Comment by jdevries3133** [comment]
# IT WORKS!!!

Wow, it finally works, that was a rough time. I think that what was actually messing me up is *this whole time* (many hours), I was trying to set a breakpoint in my `index.tsx` file, and it seems like breakpoints are working for *pretty much any other file.*

Presumably, there are a lot of `index.tsx` files, so that mapping is ambiguous or impossible –– I'm not really sure what the heck is going on there, and that's definitely something to look into further. It's also possible that I fixed my problem by slightly changing the glob pattern in the `launch.json`; this is the version that seems to be doing the trick for me now:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "address": "localhost",
      "name": "Attach to Remote",
      "port": 9229,
      "request": "attach",
      "skipFiles": ["<node_internals>/**"],
      "type": "node",
      "sourceMaps": true,
      "localRoot": "${workspaceFolder}",
      "remoteRoot": "/app",
      "outFiles": [
        "${workspaceFolder}/build/**/*.js",
        "${workspaceFolder}/public/build/**/*.js"
      ]
    }
  ]
}

```

Anyway, I am going to investigate further later to isolate the specific fix in case anyone else comes across this, but I am pretty confident that it was either a bug where `index.(jsx|tsx|js|ts)` files get messed up with source mapping, or the slight change of the glob pattern from `.../*.js` to `.../**/*.js`.

**Reply by jdevries3133** [reply]
note that breakpoints still don't work in the `index.tsx`, which is why I'm leaning more heavily towards the file that `index.tsx` files get messed up in source maps, which sounds like it could be a bug in nvim-dap or vscode-node2-debug.

**Comment by PE4CE4ALL** [comment]
Hi!
I have been trying to debug docker containers from nvim too, what I've found:
I've make it work, but I did not have many index files, and Im using vscode-js-debug on nvim  over Windows (with some problems on the path where the plugins was been installed, so I have to install it outside nvim and pass the path configuration). I've ended with this in my nvim\after\plugin\dap.lua file (Hope it can help somebody!):

-- # dap_vscode_js config --
--
require("dap-vscode-js").setup({
  -- node_path = "node", -- Path of node executable. Defaults to $NODE_PATH, and then "node"
   debugger_path = "c:\\Users\\Pablo\\dev\\microsoft\\vscode-js-debug",  -- Path to vscode-js-debug installation. -- Packer path fails: "c:\\Users\\Pablo\\AppData\\Local\\nvim-data\\site\\pack\\packer\\opt\\vscode-js-debug\\src", 
  -- debugger_cmd = { "js-debug-adapter" }, -- Command to use to launch the debug server. Takes precedence over `node_path` and `debugger_path`.
  adapters = { 'pwa-node', 'pwa-chrome', 'pwa-msedge', 'node-terminal', 'pwa-extensionHost' }, -- which adapters to register in nvim-dap
  -- log_file_path = "(stdpath cache)/dap_vscode_js.log" -- Path for file logging
  -- log_file_level = false -- Logging level for output to file. Set to false to disable file logging.
  -- log_console_level = vim.log.levels.ERROR -- Logging level for output to console. Set to false to disable console output.
})

local js_based_languages = { "typescript", "javascript",
    -- javascriptreact, typescriptreact, vue
}

--Debug docker container
local dapDockerContainerConfig = {
  name = "Docker: Attach to Node",
  type = "pwa-node",
  request = "attach",
  port = 9229,
  address = "localhost",
  localRoot = "${workspaceFolder}",
  remoteRoot = "/home/app",
  protocol = "inspector"
}

local function debugDockerContainer()
    print("debugging docker container...")
    dap.run( dapDockerContainerConfig )
end


--launch specific debugger
--some other stuff...
vim.keymap.set('n', '<leader>dd', function() debugDockerContainer() end)
